internal/revision/scanner.go

```diff
@@ -0,0 +1,124 @@
+package revision
+
+import (
+	"bufio"
+	"io"
+	"unicode"
+)
+
+var zeroRune = rune(0)
+
+// scanner represents a lexical scanner.
+type scanner struct {
+	r *bufio.Reader
+}
+
+// newScanner returns a new instance of scanner.
+func newScanner(r io.Reader) *scanner {
+	return &scanner{r: bufio.NewReader(r)}
+}
+
+// read reads the next rune from the bufferred reader.
+// Returns the rune(0) if an error occurs (or io.EOF is returned).
+func (s *scanner) read() rune {
+	ch, _, err := s.r.ReadRune()
+	if err != nil {
+		return zeroRune
+	}
+	return ch
+}
+
+// unread places the previously read rune back on the reader.
+func (s *scanner) unread() { _ = s.r.UnreadRune() }
+
+// Scan extract tokens from an input
+func (s *scanner) scan() (token, string) {
+	ch := s.read()
+
+	switch ch {
+	case zeroRune:
+		return eof, ""
+	case ':':
+		return colon, string(ch)
+	case '~':
+		return tilde, string(ch)
+	case '^':
+		return caret, string(ch)
+	case '.':
+		return dot, string(ch)
+	case '/':
+		return slash, string(ch)
+	case '{':
+		return obrace, string(ch)
+	case '}':
+		return cbrace, string(ch)
+	case '-':
+		return minus, string(ch)
+	case '@':
+		return at, string(ch)
+	case '\\':
+		return aslash, string(ch)
+	case '?':
+		return qmark, string(ch)
+	case '*':
+		return asterisk, string(ch)
+	case '[':
+		return obracket, string(ch)
+	case '!':
+		return emark, string(ch)
+	}
+
+	if unicode.IsSpace(ch) {
+		return space, string(ch)
+	}
+
+	if unicode.IsControl(ch) {
+		return control, string(ch)
+	}
+
+	if unicode.IsLetter(ch) {
+		s.unread()
+
+		return s.scanWord()
+	}
+
+	if unicode.IsNumber(ch) {
+		s.unread()
+
+		return s.scanNumber()
+	}
+
+	return char, string(ch)
```

Comment:
This needs some explanation, what is the difference between a `char` here and a `word` (from `scanWord`) and why is char the default case for this function (I would expect an error or something in the default case)?

Maybe you can answer all these question by just explaining in a comment the possible cases of your scanner at the very beginning so people understand the problem before reading your solution: Revision are very varied, and people will need to spend a lot of time reading the spec to understand all the supported types. You already spend all that time when you were writing this file so maybe you can save some people time by explaining that yourself.